romancal/outlier_detection/outlier_detection_step.py

```diff
@@ -1,37 +1,202 @@
+        pixfrac = float(default=1.0)
+        kernel = string(default='square') # drizzle kernel
+        fillval = string(default='INDEF')
+        nlow = integer(default=0)
+        nhigh = integer(default=0)
+        maskpt = float(default=0.7)
+        grow = integer(default=1)
+        snr = string(default='5.0 4.0')
+        scale = string(default='1.2 0.7')
+        backg = float(default=0.0)
+        kernel_size = string(default='7 7')
+        threshold_percent = float(default=99.8)
+        ifu_second_check = boolean(default=False)
+        save_intermediate_results = boolean(default=False)
+        resample_data = boolean(default=True)
+        good_bits = string(default="~DO_NOT_USE")  # DQ flags to allow
+        scale_detection = boolean(default=False)
+        search_output_file = boolean(default=False)
+        allowed_memory = float(default=None)  # Fraction of memory to use for the combined image
+        in_memory = boolean(default=False)
+    """
+
+    def process(self, input_models):
+        """Perform outlier detection processing on input data."""
+
+        self.input_models = input_models
+        self.input_container = isinstance(self.input_models, ModelContainer)
+        # Setup output path naming if associations are involved.
+        asn_id = None
+        with contextlib.suppress(AttributeError, KeyError):
+            asn_id = self.input_models.meta.asn_table.asn_id
+        if asn_id is None:
+            asn_id = self.search_attr("asn_id")
+        if asn_id is not None:
+            _make_output_path = self.search_attr("_make_output_path", parent_first=True)
+
+            self._make_output_path = partial(_make_output_path, asn_id=asn_id)
+
+        # Setup outlier detection parameters
+        pars = {
+            "weight_type": self.weight_type,
+            "pixfrac": self.pixfrac,
+            "kernel": self.kernel,
+            "fillval": self.fillval,
+            "nlow": self.nlow,
+            "nhigh": self.nhigh,
+            "maskpt": self.maskpt,
+            "grow": self.grow,
+            "snr": self.snr,
+            "scale": self.scale,
+            "backg": self.backg,
+            "kernel_size": self.kernel_size,
+            "threshold_percent": self.threshold_percent,
+            "ifu_second_check": self.ifu_second_check,
+            "allowed_memory": self.allowed_memory,
+            "in_memory": self.in_memory,
+            "save_intermediate_results": self.save_intermediate_results,
+            "resample_data": self.resample_data,
+            "good_bits": self.good_bits,
+            "make_output_path": self.make_output_path,
+        }
+
+        # Add logic here to select which version of OutlierDetection
+        # needs to be used depending on the input data
+        if self.input_container:
+            single_model = self.input_models[0]
+        else:
+            single_model = self.input_models
+        exptype = single_model.meta.exposure.type
+        self.check_input()
```

Comment:
This perhaps is not necessary if you remove the registry.

Reply:
Refactored the code to check for a valid `ModelContainer` input.